[FLINK-19466][FLINK-19467][runtime / state backends] Add Flip-142 public interfaces and methods

sjwiesman opened the pull request:
## What is the purpose of the change

This PR adds the new public classes as described in FLIP-142, JobManagerCheckpointStorage, FileSystemCheckpointStorage, HashMapStateBackend, and EmbeddedRocksDBStateBackend.

## Brief changelog

97e2ed8  Adds `setDefaultSavepointDir` to StreamExecutionEnvironment and wires it into the StreamConfig

1d004fab55  Adds the two checkpoint storage implementations, JobManagerCheckpointStorage and FileSystemCheckpointStorage. To maintain backward compatibility with existing flink-confs the configuration `state.checkpoint-storage` needs to be optional. The default value is `FileSystemCheckpointStorage` if a checkpoint directory is provided, `JobManagerCheckpointStorage` otherwise. I believe this will be the least surprising to users as virtually everyone who sets a checkpoint directory wants the full scalability of filesystem and not just externalized jm. Most of the tests come in the next commit where there are state backends to test these storage types with. 

791054dc  Adds the new state backends `HashMapStateBackend` and `EmbeddedRocksDBStateBackend`. Note to reviewers, I changed the existing `RocksDBStateBackend` to be a thin wrapper around the new `EmbeddedRocksDBStateBackend`. That is because it has so many configurations and is complex enough that I did not feel I could reasonably have enough test coverage over both. This way, every existing test in flink that uses RocksDB implicitly tests this new implementation so we can feel confident in its correctness. This commit also deprecates the now legacy state backends. 
 

I removed the commit for the PyFlink changes and will instead open that as a quick follow-up PR. I think it was a mistake to include that originally as it needs to be reviewed by a different set of contributors. 

## Verifying this change

*(Please pick either of the following options)*

UT / IT tests and all existing tests that use RocksDB.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented) I plan to add documentation as a follow up PR.

sjwiesman commented:
@dianfu @hequn8128 Could one of you help review the last commit?

dianfu commented:
@sjwiesman Thanks a lot for letting me know of this change. I will take a look at the last commit which contains Python changes.

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+        return CustomCheckpointStorage(j_checkpoint_storage)  # users' customized state backend
+
+
+class CheckpointStorage(object):
```
```suggestion
class CheckpointStorage(abc.ABC):
```

See https://issues.apache.org/jira/browse/FLINK-19892 for more details.

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+    streams concurrently.
+    """
+
+    __metaclass__ = ABCMeta
```
We can remove this then.

dianfu commented on flink-python/pyflink/datastream/checkpoint_config.py:
```diff
@@ -330,6 +331,47 @@ def disable_unaligned_checkpoints(self) -> 'CheckpointConfig':
+        :param checkpoint_storage: The :class:`CheckpointStorage`.
+        """
+
+        self._j_checkpoint_config.setCheckpointStorage(checkpoint_storage._j_checkpoint_storage)
```
```suggestion
        self._j_checkpoint_config.setCheckpointStorage(checkpoint_storage._j_checkpoint_storage)
        return self
```

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+        self._j_jobmanager_checkpoint_storage = j_jobmanager_checkpoint_storage
+        super(CheckpointStorage, self).__init__(j_jobmanager_checkpoint_storage)
+
+    def get_max_state_size(self):
```
```suggestion
    def get_max_state_size(self) -> int:
```

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+    DEFAULT_MAX_STATE_SIZE = 5 * 1024 * 1024
+
+    def __init__(self,
+                 checkpoint_path=None,
```
Add type hint for these parameters?

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+    """
+
+    def __init__(self,
+                 checkpoint_directory_uri=None,
```
Add typehint for these parameters?

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+            j_jobmanager_checkpoint_storage = JJobManagerCheckpointStorage(checkpoint_path,
+                                                                           max_state_size)
+
+        self._j_jobmanager_checkpoint_storage = j_jobmanager_checkpoint_storage
```
could be removed

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+                                                                   file_state_size_threshold,
+                                                                   write_buffer_size)
+
+        self._j_fs_checkpoint_storage = j_fs_checkpoint_storage
```
could be removed

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+        self._j_fs_checkpoint_storage = j_fs_checkpoint_storage
+        super(FileSystemCheckpointStorage, self).__init__(j_fs_checkpoint_storage)
+
+    def get_checkpoint_path(self):
```
```suggestion
    def get_checkpoint_path(self) -> str:
```

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+        """
+        return self._j_fs_checkpoint_storage.getCheckpointPath().toString()
+
+    def get_min_file_size_threshold(self):
```
```suggestion
    def get_min_file_size_threshold(self) -> int:
```

dianfu commented on flink-python/pyflink/datastream/checkpoint_storage.py:
```diff
@@ -0,0 +1,302 @@
+        """
+        return self._j_fs_checkpoint_storage.getMinFileSizeThreshold()
+
+    def get_write_buffer_size(self):
```
```suggestion
    def get_write_buffer_size(self) -> int:
```

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+                                "True or False.")
+            j_hashmap_state_backend = JHashMapStateBackend(j_asynchronous_snapshots)
+
+        self._j_memory_state_backend = j_hashmap_state_backend
```
could be removed

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+        self._j_memory_state_backend = j_hashmap_state_backend
+        super(HashMapStateBackend, self).__init__(j_hashmap_state_backend)
+
+    def is_using_asynchronous_snapshots(self):
```
```suggestion
    def is_using_asynchronous_snapshots(self) -> bool:
```

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+            j_rocks_db_state_backend = \
+                JEmbeddedRocksDBStateBackend(j_enable_incremental_checkpointing)
+
+        self._j_rocks_db_state_backend = j_rocks_db_state_backend
```
could be removed

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+        self._j_rocks_db_state_backend = j_rocks_db_state_backend
+        super(EmbeddedRocksDBStateBackend, self).__init__(j_rocks_db_state_backend)
+
+    def set_db_storage_paths(self, *paths):
```
```suggestion
    def set_db_storage_paths(self, *paths: str):
```

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+                j_path_array[i] = paths[i]
+            self._j_rocks_db_state_backend.setDbStoragePaths(j_path_array)
+
+    def get_db_storage_paths(self):
```
```suggestion
    def get_db_storage_paths(self) -> str:
```

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+        """
+        return list(self._j_rocks_db_state_backend.getDbStoragePaths())
+
+    def is_incremental_checkpoints_enabled(self):
```
```suggestion
    def is_incremental_checkpoints_enabled(self) -> bool:
```

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+        """
+        return self._j_rocks_db_state_backend.isIncrementalCheckpointsEnabled()
+
+    def set_predefined_options(self, options):
```
```suggestion
    def set_predefined_options(self, options: PredefinedOptions):
```

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+
+        :param options: The options to set (must not be null), see :class:`PredefinedOptions`.
+        """
+        gateway = get_gateway()
```
could be replaced with only one line as following:
```
self._j_rocks_db_state_backend.setPredefinedOptions(options._to_j_predefined_options()
```

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+
+        :return: Current predefined options.
+        """
+        j_predefined_options = self._j_rocks_db_state_backend.getPredefinedOptions()
```
could be optimized as following:
```       
 j_predefined_options = self._j_rocks_db_state_backend.getPredefinedOptions()
return PredefinedOptions._from_j_predefined_options(j_predefined_options)
```

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+                            " PredefinedOptions.SPINNING_DISK_OPTIMIZED_HIGH_MEM and "
+                            "PredefinedOptions.FLASH_SSD_OPTIMIZED")
+
+    def get_predefined_options(self):
```
add type hint

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+        else:
+            raise Exception("Unsupported java options: %s" % j_predefined_options)
+
+    def set_options(self, options_factory_class_name):
```
add type hint for parameter options_factory_class_name

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+            raise ValueError("The input class does not implement RocksDBOptionsFactory.")
+        self._j_rocks_db_state_backend.setRocksDBOptions(j_options_factory_clz.newInstance())
+
+    def get_options(self):
```
add typehint

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+        else:
+            return None
+
+    def get_number_of_transfering_threads(self):
```
ditto

dianfu commented on flink-python/pyflink/datastream/state_backend.py:
```diff
@@ -116,8 +125,320 @@ def __init__(self, j_state_backend):
+        """
+        return self._j_rocks_db_state_backend.getNumberOfTransferingThreads()
+
+    def set_number_of_transfering_threads(self, number_of_transfering_threads):
```
ditto

dianfu commented on flink-python/pyflink/datastream/stream_execution_environment.py:
```diff
@@ -280,7 +276,38 @@ def set_state_backend(self, state_backend: StateBackend) -> 'StreamExecutionEnvi
         return self
 
-    def set_restart_strategy(self, restart_strategy_configuration: RestartStrategyConfiguration):
+    def set_default_savepoint_directory(self, savepoint_path):
```
add type hints

dianfu commented on flink-python/pyflink/datastream/stream_execution_environment.py:
```diff
@@ -280,7 +276,38 @@ def set_state_backend(self, state_backend: StateBackend) -> 'StreamExecutionEnvi
+            self._j_stream_execution_environment.setDefaultSavepointDirectory(savepoint_path)
+        return self
+
+    def get_default_savepoint_directory(self):
```
ditto

dianfu reviewed (commented):
@sjwiesman Thanks for the PR. I have left a few comments about the last commit. Most of them about adding type hints which was introduced recently in FLINK-16522.

sjwiesman commented:
Thanks for the review @dianfu . I've added typehints to all the methods and classes along with applying the simplifications you suggested.

dianfu commented on flink-python/pyflink/datastream/tests/test_check_point_config.py:
```diff
@@ -151,3 +152,12 @@ def test_is_unaligned_checkpointing_enabled(self):
+
+    def test_set_checkpoint_storage(self):
+
+        self.assertNone(self.checkpoint_config.get_checkpoint_storage())
```
```suggestion
        self.assertIsNone(self.checkpoint_config.get_checkpoint_storage())
```

dianfu reviewed (commented):
@sjwiesman Thanks a lot for the update. Have left a few more comments, most of them are about the checkstyle.

sjwiesman commented:
@flinkbot run azure

sjwiesman commented:
@flinkbot run azure

sjwiesman commented:
@flinkbot run azure

sjwiesman commented:
@flinkbot run azure

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/storage/FileSystemCheckpointStorage.java:
```diff
@@ -0,0 +1,322 @@
+    private FileSystemCheckpointStorage(
+            FileSystemCheckpointStorage original,
+            ReadableConfig configuration,
+            ClassLoader classLoader) {
```
`classLoader` is never used.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/storage/JobManagerCheckpointStorage.java:
```diff
@@ -0,0 +1,218 @@
+ * implemented via the {@link #configure(ReadableConfig, ClassLoader)} method.
+ */
+@PublicEvolving
+public class JobManagerCheckpointStorage extends AbstractFileCheckpointStorage
+        implements CheckpointStorage, ConfigurableCheckpointStorage {
```
I don't think that inheritance is necessary here. After moving two static methods from `AbstractFileCheckpointStorage`, only two fields an a call to static  `AbstractFsCheckpointStorageAccess.resolveCheckpointPointer` are inherited.

Having JM storage extending FS storage is confusing to me, as TM doesn't actually use FS.

sjwiesman replied:
That's fair, I'll rework this.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/storage/JobManagerCheckpointStorageFactory.java:
```diff
@@ -0,0 +1,41 @@
+        implements CheckpointStorageFactory<JobManagerCheckpointStorage> {
+
+    @Override
+    public JobManagerCheckpointStorage createFromConfig(
```
How about moving this method to `JobManagerCheckpointStorage` and making it static?
(and removing this class)

ditto: `FileSystemCheckpointStorageFactory`

rkhachatryan reviewed (commented):
Thanks for the contribution @sjwiesman 
I've made the first pass through the PR. In general it LGTM.
I'll take a look at other parts soon.

sjwiesman commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java:
```diff
@@ -84,12 +88,23 @@
             case JOB_MANAGER_STORAGE_NAME:
-                throw new UnsupportedOperationException(
-                        "JobManagerCheckpointStorage is not yet implemented");
+                if (logger != null) {
```
I copied this pattern from StateBackendLoader without checking why. Let me see if there's a good reason for it, otherwise +1

sjwiesman commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/storage/AbstractFileCheckpointStorage.java:
```diff
@@ -0,0 +1,215 @@
+     * @return The URI as a Path.
+     * @throws IllegalArgumentException Thrown, if the URI misses scheme or path.
+     */
+    protected static Path validatePath(Path path) {
```
I'm not sure, some of these are fairly checkpointing specific but I do think this can be cleaned up and less low-level.

sjwiesman commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/storage/AbstractFileCheckpointStorage.java:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    @Nullable
+    protected static Path parameterOrConfigured(
```
The `Function<String, T>` seems wrong to me. I think we can instead add `Path` as a first class config type.

sjwiesman replied:
Making Path a first class config type ended up being way too involved. I did clean this up.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackendFactory.java:
```diff
@@ -0,0 +1,33 @@
+import org.apache.flink.runtime.state.StateBackendFactory;
+
+/** A factory that creates an {@link EmbeddedRocksDBStateBackend} from a configuration. */
+public class EmbeddedRocksDBStateBackendFactory
```
Is it `@PublicEvolving`?

ditto `HashMapStateBackendFactory`
and `HashMapStateBackend`

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -103,6 +108,14 @@ public static StateBackend loadStateBackendFromConfig(
+                    logger.info("State backend is set to heap memory {}", hashMapStateBackend);
+                }
+                return hashMapStateBackend;
+
             case MEMORY_STATE_BACKEND_NAME:
                 MemoryStateBackend memBackend =
                         new MemoryStateBackendFactory().createFromConfig(config, classLoader);
```
Why do we still need to create and return deprecated `MemoryStateBackend`? Why can't we return `HashMapStateBackend` as above?

ditto `FS_STATE_BACKEND_NAME`

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java:
```diff
@@ -131,7 +144,7 @@ public static StateBackend loadStateBackendFromConfig(
 
             case ROCKSDB_STATE_BACKEND_NAME:
                 factoryClassName =
-                        "org.apache.flink.contrib.streaming.state.RocksDBStateBackendFactory";
+                        "org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackendFactory";
```
I think this change breaks backwards combatibility (with old configs). 
Should we fallback to the older factory/name?

sjwiesman replied:
No, it doesn't. If you do not explicitly specify a checkpoint storage policy and a checkpoint directory is provided it will use FileSystemCheckpointStorage which is equivalent to the old RocksDBStateBackend.

rkhachatryan replied:
You're right, thanks. I misread this part of code.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:
```diff
@@ -562,27 +566,30 @@ public CheckpointingMode getCheckpointingMode() {
-     * Sets the state backend that describes how to store and checkpoint operator state. It defines
-     * both which data structures hold state during execution (for example hash tables, RockDB, or
-     * other data stores) as well as where checkpointed data will be persisted.
+     * Sets the state backend that describes how to store operator. It defines the data structures
+     * that hold state during execution (for example hash tables, RocksDB, or other data stores).
```
I think there is a mistake in this change: shouldn't it be "how to checkpoint operator state"?

ditto: `StreamExecutionEnvironment`

sjwiesman replied:
From a user perspecitve the state backend stores state locally while checkpoint storage defines how it is checkpointed. I understand under the hood its fuzzier than that but I think this javadoc is correct.

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/CheckpointingOptions.java:
```diff
@@ -27,22 +29,70 @@
     // ------------------------------------------------------------------------
 
-    /** The state backend to be used to store and checkpoint state. */
+    /**
+     * The checkpoint storage used to store state.
+     *
+     * <p>The implementation can be specified either via their shortcut name, or via the class name
+     * of a {@code StateBackendFactory}. If a StateBackendFactory class name is specified, the
+     * factory is instantiated (via its zero-argument constructor) and its {@code
+     * StateBackendFactory#createFromConfig(ReadableConfig, ClassLoader)} method is called.
+     *
+     * <p>Recognized shortcut names are 'hashmap' and 'rocksdb'.
+     */
```
I guess this change and the one below were made by mistake. This is actually about `CHECKPOINT_STORAGE`, right?

sjwiesman replied:
I've expanded on the javadoc but this is about the statebackend.

rkhachatryan commented on docs/_includes/generated/checkpointing_configuration.html:
```diff
@@ -60,7 +54,19 @@
+            <td>The default directory for savepoints. Used by the state backends that write savepoints to file systems (HashMapStateBackend, EmbeddedRocksDBStateBackend).</td>
+        </tr>
+        <tr>
+            <td><h5>state.snapshot.fs.memory-threshold</h5></td>
```
How about `state.storage.fs.memory-threshold` instead of  `state.snapshot.fs.memory-threshold`?

ditto `write-buffer-size`

rkhachatryan reviewed (changes requested):
Thanks for updating your PR @sjwiesman 
I finished my review and left some comments, PTAL.
(and I think a couple of questions from the previous round are still unresolved).

sjwiesman commented:
@rkhachatryan I believe I have addressed all your comments. I'm signing off tonight before seeing if the e2e tests pass but if anything fails I will fix it first thing in the morning.

rkhachatryan commented:
Thanks for updating the PR @sjwiesman. 
The changes LGTM, but I see build failures that are likely related (`StateBackendLoadingTest`).
Some minor commits are not addressed yet ([1](https://github.com/apache/flink/pull/13912#discussion_r566225179), [2](https://github.com/apache/flink/pull/13912#discussion_r566271838)). They are not very important so please feel free to skip them.

sjwiesman commented:
@flinkbot run azure

sjwiesman commented:
@flinkbot run azure
